SCC sorting by priority

pweil- opened the pull request:
Sorts SCCs by a priority field and then by name.  Adds a priority to the privileged SCC so that if you have administrative access you will use that SCC by default (though this is largely unnecessary since the first commit allows authenticated users to run as any uid).

Builds on https://github.com/openshift/origin/pull/5498 which is hanging out in the merge queue right now so only the last three commits need reviewed.

Fixes https://github.com/openshift/origin/issues/5317

@liggitt @smarterclayton

pweil- commented:
fixing commits so I don't incur rebase wrath, I packaged v1beta3 in with non-v1beta3.

liggitt commented:
Ones with priority specified should come before ones without. Equal priority should maybe sort by the current sort behavior, rather than name, for backwards compatibility

liggitt commented:
In the case of a tie between priority AND old security score, name is fine to break the tie

liggitt commented on Godeps/_workspace/src/k8s.io/kubernetes/pkg/api/v1/types.go:
```diff
@@ -2590,6 +2590,11 @@ type SecurityContextConstraints struct {
+	// SortPriority influences the sort order of SCCs when evaluating which SCCs to try first for
+	// a given pod request based on access in the Users and Groups fields.  The higher the int, the
+	// higher priority.  If scores for multiple SCCs are equal they will be sorted by name.
+	SortPriority int `json:"sortPriority" description:"influences the sort order when evaluating SCCs available to the user"`
```
"determines which SCC is used when multiple SCCs allow a particular pod; higher priority SCCs are preferred"

liggitt replied:
I would prefer `priority` over `sortPriority`... priority implies a sort, I think

liggitt commented on pkg/cmd/server/bootstrappolicy/securitycontextconstraints.go:
```diff
@@ -69,11 +221,54 @@ func GetBootstrapSecurityContextConstraints(buildControllerUsername string) []ka
 			},
-			Groups: []string{AuthenticatedGroup},
 		},
+		// SecurityContextConstraintsAnyUID allows no host access and allocates SELinux.
```
I thought this was the one we wanted admins to default to... selinux, but not forcing a uid

smarterclayton replied:
Only once 1.9 docker lands.  Probably will be optional and user driven.

On Oct 30, 2015, at 3:32 PM, Jordan Liggitt <notifications@github.com>
wrote:

In pkg/cmd/server/bootstrappolicy/securitycontextconstraints.go
<https://github.com/openshift/origin/pull/5543#discussion_r43543249>:

>  		},
> +		// SecurityContextConstraintsAnyUID allows no host access and allocates SELinux.

I thought this was the one we wanted admins to default to... selinux, but
not forcing a uid

—
Reply to this email directly or view it on GitHub
<https://github.com/openshift/origin/pull/5543/files#r43543249>.

pweil- replied:
Ok, to make sure I'm clear on addressing this comment.  We have the following SCCs now:

1. privileged - all host access, priv access, RunAsAny for all strats
1. restricted - no host access, requires uid/selinux/fsgroup/sup grups
1. nonroot - same as restricted but with the NonRoot strategy for uid
1. hostmount - same as restricted but allows host mounts
1. hostaccess - same as restricted but allows access to all host ns features
1. anyuid - no host features, no priv, runAsAny for all strats except SELinux

Right now I've defaulted the authenticated group to anyuid and given a priority of 1 to privileged.  So all users have access to anyuid and, if you have access to privileged, you'll use it.

Is this incorrect?

pweil- replied:
discussed in IRC...fixing

liggitt replied:
I think this is the summary:
* `system:authenticated` should remain in the `restricted` SCC for now
* `system:cluster-admins` should be in `privileged` and `anyuid`
* `anyuid` should have `priority` of `1`, others should have `priority` of `0`. This means anyone with access to the `anyuid` SCC will not have uids assigned to their pods automatically.

@smarterclayton does that sound right? note that the `anyuid` SCC would only apply to pods the cluster admin created directly, not ones created via rc or dc. If they wanted those to use `anyuid`, they'd have to add the service account(s) they wanted to run as any uid to the SCC

liggitt commented on pkg/security/admission/byrestrictions.go:
```diff
@@ -1,52 +0,0 @@
-package admission
```
I would keep this for backwards compatibility and make it the secondary sort after priority order

deads2k commented:
I was wondering this during SCC review as well.  Why not simply merge all the user's available SCCs together, taking the highest privilege and preserving ranges alongside the "any's" for defaulting purposes.  Then attempt to fill in the default pod a single time instead of once per SCC with the full set of available options.  Wouldn't that solve the union problem, the separate of SCCs without specifying all combinations problem, the improper priority problem, and make the code easier to understand all at the same time?  Seems like that would only become weird if you wanted crazy things like, "you can use these GIDs, but only as this UID".

I'm guessing you guys considered and discarded this, but I'd like to understand why.

liggitt commented:
> Why not simply merge all the user's available SCCs together, taking the highest privilege and preserving ranges alongside the "any's" for defaulting purposes. Then attempt to fill in the default pod a single time instead of once per SCC with the full set of available options. Wouldn't that solve the union problem, the separate of SCCs without specifying all combinations problem, the improper priority problem, and make the code easier to understand all at the same time?

Mixing and matching bits of different SCCs isn't a good idea... privileges can combine to allow things that weren't intended. For example:
* SCC A lets you host mount, but only as a particular user and with a specific selinux context
* SCC B does not let you host mount, but lets you run as any non-root user

Mixing and matching would let someone host mount without getting the accompanying uid and selinux restriction the admin specified.

deads2k commented:
Maybe there's a better structure to use for constraints or the bindings.  Having to specify every possible combination is pretty ridiculous and scales horribly as more options are added.  Even just assuming on/off, we have 11 options making 2048 different potential policies to manage instead of 11.

liggitt commented:
> 1.6x10^5894 options.

Sounds pretty powerful to me :)

smarterclayton commented:
I doubt you're going to see more than 10 in use at any one site.  And we
can survive until then.

On Oct 31, 2015, at 8:25 AM, David Eads <notifications@github.com> wrote:

Maybe there's a better structure to use for constraints or the bindings.
Having to specify every possible combination is pretty ridiculous and
scales horribly as more options are added. Even just assuming on/off, we
have 11 options making 2048 different potential policies to manage instead
of 11.

—
Reply to this email directly or view it on GitHub
<https://github.com/openshift/origin/pull/5543#issuecomment-152732069>.

pweil- commented:
@liggitt - round 2 ready for your eyes

1.  sort by priority first
1.  if priorities are equal, sort by score in the old behavior
1.  if priorities and score are equal sort by name
1.  admin has access to privileged and anyuid
1.  anyuid is given priority 1
1.  changed field and descriptions according to comments

liggitt commented on Godeps/_workspace/src/k8s.io/kubernetes/pkg/api/v1/types.go:
```diff
@@ -2590,6 +2590,11 @@ type SecurityContextConstraints struct {
+	// Priority influences the sort order of SCCs when evaluating which SCCs to try first for
+	// a given pod request based on access in the Users and Groups fields.  The higher the int, the
+	// higher priority.  If scores for multiple SCCs are equal they will be sorted by name.
+	Priority int `json:"sortPriority" description:"determines which SCC is used when multiple SCCs allow a particular pod; higher priority SCCs are preferred"`
```
change json field name to `priority`

pweil- replied:
Damn, you're too quick.  I just caught it.  Fixed.

liggitt commented on Godeps/_workspace/src/k8s.io/kubernetes/pkg/api/v1beta3/types.go:
```diff
@@ -2085,6 +2085,11 @@ type SecurityContextConstraints struct {
+	// Priority influences the sort order of SCCs when evaluating which SCCs to try first for
+	// a given pod request based on access in the Users and Groups fields.  The higher the int, the
+	// higher priority.  If scores for multiple SCCs are equal they will be sorted by name.
+	Priority int `json:"sortPriority" description:"determines which SCC is used when multiple SCCs allow a particular pod; higher priority SCCs are preferred"`
```
change json field name to `priority`

liggitt commented on api/swagger-spec/api-v1.json:
```diff
@@ -14761,6 +14762,11 @@
      "metadata": {
       "$ref": "v1.ObjectMeta"
      },
+     "sortPriority": {
```
regen, should be `priority`

pweil- replied:
fixed

liggitt commented on pkg/cmd/server/bootstrappolicy/securitycontextconstraints.go:
```diff
@@ -246,6 +246,8 @@ func GetBootstrapSecurityContextConstraints(sccNameToAdditionalGroups map[string
 			SupplementalGroups: kapi.SupplementalGroupsStrategyOptions{
 				Type: kapi.SupplementalGroupsStrategyRunAsAny,
 			},
+			// by default if you have access to the admin strategy then use it
```
to the `anyuid` strategy`

pweil- replied:
fixed

liggitt commented on pkg/security/admission/admission_test.go:
```diff
@@ -68,9 +68,11 @@ func TestAdmit(t *testing.T) {
 		},
 		Groups: []string{"system:serviceaccounts"},
+		// give this scc priority since it is what we want to validate with
+		SortPriority: 1,
```
don't you need a more restrictive SCC that also matches to actually check that you're matching on sort priority, and not just the secondary `pointScore` sort?

pweil- replied:
actually, I don't need this priority at all.  The SCC below it is more restrictive (a MustRunAs uid strategy rather than a MustRunAsRange) that is in `scc-sa`.  Removing the priority they would have equal priorities.  The scores should ensure that `scc-sa-exact` is tried before `scc-sa` but it should be failing to validate.  We check the matched name in the test run which will ensure we're matching against what we expect....removing.

liggitt commented:
LGTM

liggitt commented:
[test]

pweil- commented:
re[test]

pweil- commented:
[test]

pweil- commented:
moar [test]!

liggitt commented:
`'[ "$(oc rsh ${BUILD_ID}-build 2>&1 | grep 'forbidden')" ]' exited with status 1` fixes, re[test]

pweil- commented:
woohoo!

liggitt commented:
thinking through the reconcile command... do we want priority to be an *int so we can tell the difference between priority of `0` and unset?

pweil- commented:
Looking at priorities when listing SCCs you can see, at a glance, which would be favored (based on priority only, score is not shown).  Throwing nils into the mix might be confusing, especially if you consider -1 > nil if someone wanted to use a negative to throw an SCC to the back of the line.  

Though that does make me think that we should show them in a default sort order when listing (post 3.1).  That would make it much nicer.

liggitt commented:
So when reconciling, how do you know when a priority has been set to zero by the cluster Edmond and we shouldn't touch it, or it hasn't been set, and we should default?

pweil- commented:
if the flag for reconciling priorities is set true then use the value from bootstrap.  If it is set to false then  use the value on the current SCC rather than the bootstrap SCC.   I don't think what it's set to matters when detecting a diff between the bootstrap scc and current SCCs only is it different if we're reconciling the field.

liggitt commented:
it matters when we want to set defaults if the admin hasn't expressed a priority intent. Also, validate priority to be `>= 0`.

liggitt commented:
I would expect the following behavior:

Missing priority behaves like priority=0 for purposes of sorting
Missing priority gets set on reconcile
Present priority only gets set on reconcile if they explicitly do `--overwrite-priority` or similar

pweil- commented:
> Missing priority gets set on reconcile
> Present priority only gets set on reconcile if they explicitly do --overwrite-priority or similar

I see what you're saying here.  I was considering it an option of either "overwrite all priorities" or "don't touch priorities"  during reconcile but if the options are really "set missing" or "overwrite all" then nil makes sense.  That was my disconnect.

In that case, the first reconcile run (regardless of option) will set all the missing priorities on bootstrapped commands and subsequent changes to the bootstrap SCC priorities will never update until `--overwrite-priority` is used.  Sound about right to you?

liggitt commented:
> In that case, the first reconcile run (regardless of option) will set all the missing priorities on bootstrapped commands and subsequent changes to the bootstrap SCC priorities will never update until --overwrite-priority is used. Sound about right to you? 

Yes, so choose priorities carefully (or just set the one we care about to 1?)

pweil- commented:
@liggitt - updated to be a pointer, added checks in sorting, and a nil vs non-nil test.  Also updated the printer so that nil priority doesn't show as 0 which would be confusing since 0 > nil now.  Please take another pass.

pweil- commented:
oh, btw, I set anyuid to 10 to give wiggle room if an admin wants to adjust other priorities (so they don't have to immediately adjust anyuid just to keep it prioritized).

liggitt commented:
0 should equal nil for sorting purposes, imo

liggitt commented:
add validation to make sure non-nil priority is >= 0

liggitt commented on pkg/cmd/server/bootstrappolicy/securitycontextconstraints.go:
```diff
@@ -39,6 +39,14 @@ const (
+	var (
+		// this is set to 10 to allow wiggle room for admins to set other priorities without
+		// having to adjust anyUID.
+		securityContextConstraintsAnyUIDPriority = 10
```
```
10 PRINT "HELLO"
20 PRINT "WORLD"
```

pweil- replied:
parse exception.  Do you want the lower score (except 0) to be considered higher priority?

liggitt replied:
I was just amused at setting to 10 to allow space... took me back...

liggitt commented on pkg/security/admission/bypriority.go:
```diff
@@ -0,0 +1,57 @@
+func (s ByPriority) Less(i, j int) bool {
+	iSCC := s[i]
+	jSCC := s[j]
+
```
I would just do
```
iSCCPriority := 0
if iSCCPriority.Priority != nil {
  iSCCPriority = *iSCCPriority.Priority
}

jSCCPriority := 0
if jSCCPriority.Priority != nil {
  jSCCPriority = *jSCCPriority.Priority
}
```

I think unset and 0 should match in terms of priority

liggitt commented on Godeps/_workspace/src/k8s.io/kubernetes/pkg/kubectl/resource_printer.go:
```diff
@@ -1430,9 +1431,14 @@ func printHorizontalPodAutoscalerList(list *extensions.HorizontalPodAutoscalerLi
 
 func printSecurityContextConstraints(item *api.SecurityContextConstraints, w io.Writer, withNamespace, wide, showAll bool, columnLabels []string) error {
-	_, err := fmt.Fprintf(w, "%s\t%t\t%v\t%t\t%s\t%s\t%s\t%s\n", item.Name, item.AllowPrivilegedContainer,
+	priority := ""
```
`"unset"` over empty string?

liggitt replied:
what do other describers do for unset values?

liggitt replied:
oh, yeah... add to describer?

pweil- replied:
looks like "<none>" is the default.  There is no describer for SCCs yet, issue is captured in https://github.com/openshift/origin/issues/5299

pweil- commented:
made nil == 0 for sorting, added <none> in the printer, added validation method for negative priorities.

liggitt commented:
LGTM

smarterclayton commented:
[test]
